### ISSUE TITLE
Authentication doesn't work with prefix? [BUG] 

### ISSUE THREAD

**Issue by alexanfl** [bokeh/bokeh#11991]
### Software versions

Python version      :  3.8.12 | packaged by conda-forge | (default, Sep 29 2021, 19:14:40) [MSC v.1916 64 bit (AMD64)]
IPython version     :  7.28.0
Tornado version     :  6.1
Bokeh version       :  2.4.0
BokehJS static path :  C:\Users\...\Miniconda3\envs\...\lib\site-packages\bokeh\server\static
node.js version     :  v14.14.0
npm version         :  7.11.2

### Expected behavior

I expected that login works when I supply a prefix (e.g. when using `bokeh serve` or initializing `Server` with a prefix).

### Observed behavior

It seems that the `login_url` variable isn't passed _with_ the prefix to Tornado when the initial redirect to the login endpoint occurs. 

This is what happens:

- I try to go to the index page of the app (http://localhost:5006/pre/app)
- I am redirected to http://localhost:5006**/login**?next=%2Fpre%2Fapp, but it should be http://localhost:5006**/pre/login**?next=%2Fpre%2Fapp.
- If I go to the correct url, it works, but I get another 404 when I login and get redirected from there.
- If I (just to test) change the `login_url` to include the prefix, the initial redirect "works", but I don't reach the Bokeh endpoints since they will be on the form http://localhost:5006**/pre/pre/login/**.

In my actual code, I have created a work around by editing `login_url` after initializing the Server object:

```python
auth_provider = AuthModule('/path/to/module.py')
app = build_single_handler_application(full_path)
app = Server(
    {'/': app},
    prefix=prefix,
    auth_provider=auth_provider)
auth_provider._module.login_url = f'/{prefix}/login/'
```

### Example code

```Python
Run `bokeh serve app.py --auth-module=auth.py --prefix=pre`


# app.py

from bokeh.io import curdoc
from bokeh.models import Div


def modify_doc(doc):
    doc.title = 'Hello World'
    header = Div(text=f'<h1>{doc.title}</h1>')
    doc.add_root(header)


modify_doc(curdoc())



```python
# auth.py
from tornado.web import RequestHandler


login_url = '/login'


class LoginHandler(RequestHandler):
    def get(self):
        self.write('<html><body><form action="/login" method="post">'
                   'Name: <input type="text" name="name">'
                   '<input type="submit" value="Sign in">'
                   '</form></body></html>')

    def post(self):
        name = self.get_argument('name')
        self.set_cookie('user', name)
        self.redirect('/')


def get_user(request_handler):
    return request_handler.get_cookie('user')
```
```


### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

**Comment by alexanfl** [comment]
@bryevdv Yes, let me see what I can do.

**Comment by alexanfl** [comment]
@bryevdv I was thinking of updating the `get_login_url` method of `AuthRequestHandler` to add the prefix to the returned url like this:
```python
def get_login_url(self)"
    prefix = getattr(self.application, 'prefix', '')
    if self.application.auth_provider.get_login_url is not None:
        return prefix + self.application.auth_provider.get_login_url(self)
    if self.application.auth_provider.login_url is not None:
        return prefix + self.application.auth_provider.login_url
```
If this is an ok solution, do you have any suggestions on how to test this? I can't seem to be able to instantiate the class in a simple way.

**Comment by bh-tt** [comment]
@alexanfl Have you had any time to look at this?

**Comment by alexanfl** [comment]
@bh-tt, @bryevdv Sorry for the late reply. I haven't looked at this since the suggested solution. But I spent some time then familiarizing myself with the different parts of the test suite, and `test_auth_provider.py` was one of the files I looked the most at without luck. I will take another look and get back to you with specific problems I have. Hopefully I can do it this week.

@bryevdv I'll do something to ensure appropriate slashes.

**Comment by kithuto** [comment]
Hi @alexanfl . The problem is that the automatic redirect made by bokeh ignores the app prefix.
The only solution I found to this issue is using nginx and redirecting the login url to the correct route (this example is using nginx as a proxy).

Simple example of nginx configuration file:

server {
        server_name my.domain.com;
        listen 80; (or other port you can use 443 ssl and certificate too)

        location /login {
                return 301 https://$http_host{prefix in your bokeh app}$request_uri?$args;
        }

        location {prefix in your bokeh app} {
                proxy_pass http://path_to_your_bokeh_app;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
                proxy_http_version 1.1;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header Host $host:$server_port;
                proxy_buffering off;
        }
}

If anyone find a better solution I'll be happy to hear it. (I'm using LoginHandler too)

**Comment by kithuto** [comment]
@alexanfl The only solution I found is using nginx and fixing the redirection there but right now bokeh still fails to redirect correcty using LoginHandler class.... Your proposed solution will fix this problem with the LoginHandler class?

**Comment by alexanfl** [comment]
@kithuto Yes, it should, but I unfortunately completely forget about this. 

In the project I was working on where this was a problem, I solved it with a temporary workaround (which I have tried to make generic for this post):

```python
        auth_provider = AuthModule('/path/to/login/module')  # from bokeh.server.auth_provider import AuthModule
        app = build_single_handler_application(...)  # from bokeh.command.util import build_single_handler_application

        app_kwargs = {
            'port': port,
            'prefix': prefix,
            'auth_provider': auth_provider,
        }

        app = Server({'/': app}, **app_kwargs)  # from bokeh.server.server import Server
        
        # Due to a bug in Bokeh, we must add the prefix to auth_provider._module
        if isinstance(auth_provider, AuthModule) and not auth_provider._module.login_url.startswith(prefix):
            auth_provider._module.login_url = f'{prefix}/{AUTH_ENDPOINT}/'
```

**Comment by shoaib42** [comment]
After banging my head for hours, I stumbled across this. Will try to use this, but would be nice if I could just use `bokeh serve`
